build/accelerate_launch.py

```diff
@@ -117,18 +128,91 @@ def main():
             sys.exit(INTERNAL_ERROR_EXIT_CODE)
 
         try:
+            last_checkpoint_dir = get_highest_checkpoint(tempdir)
+            last_checkpoint_path = os.path.join(tempdir, last_checkpoint_dir)
+
+            use_flash_attn = job_config.get("use_flash_attn", True)
+            adapter_config_path = f"{last_checkpoint_path}/adapter_config.json"
+            tokenizer = AutoTokenizer.from_pretrained(last_checkpoint_path)
+
+            if os.path.exists(adapter_config_path):
+                base_model_path = get_base_model_from_adapter_config(
+                    adapter_config_path
+                )
+                base_model = AutoModelForCausalLM.from_pretrained(
+                    base_model_path,
+                    attn_implementation="flash_attention_2" if use_flash_attn else None,
+                    torch_dtype=bfloat16 if use_flash_attn else None,
+                )
+
+                # since the peft library (PEFTModelForCausalLM) does not handle cases
+                # where the model's layers are modified, in our case the embedding layer
+                # is modified, so we resize the backbone model's embedding layer with our own
+                # utility before passing it along to load the PEFT model.
+                tokenizer_data_utils.tokenizer_and_embedding_resize(
```

Comment:
is there a reason that this needs to happen again? I think this is called by the launched sft trainer command, so shouldn't the resizing already be handled?

Reply:
Unit tests for LoRA and PT were failing when trying to load the checkpoint with error:
```
RuntimeError: Error(s) in loading state_dict for LlamaForCausalLM:
	size mismatch for model.embed_tokens.weight: copying a param with shape torch.Size([32008, 64]) from checkpoint, the shape in current model is torch.Size([32000, 64]).
```
so then I had to resize the tokens to be able to load the checkpoint.

Reply:
Iiiiinteresting. I guess the resize information must not be preserved in the config and reloading with the base class resets it to the base size or something 🤔

Reply:
I also verified that I get this same error when manually running tuning on tiny llama model with same params as unit test and then trying to load the checkpoint.

Reply:
huh? is this happening on current main due to some other changes in dependencies or something? 

why is this change related to the lm_head deletion here?

Reply:
if something changed, we should not just patch it by doing such things for final checkpoint only. Though product uses final checkpoint only for now , rest of research can use any checkpoint . so our load and inference scripts should work with any checkpoint

Reply:
I didn't think this was a patch, I thought this was always needed to load a model we tuned such as what we run in [run_inference](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/scripts/run_inference.py#L193-L198). Since we update the tokenizer in our sft_trainer then when we load the tuned checkpoint again we need to update the tokenizer as well

Reply:
I didn't realize this was a new change added in a few weeks ago - https://github.com/foundation-model-stack/fms-hf-tuning/pull/227/files -- that resizes the embeddings to be a multiple of 8 with new tokens instead of the size of the tokenizer which required this addition....hmmm